### ISSUE TITLE
Update to v14 breaks @testing-library/user-event on Vitest

### ISSUE THREAD

**Issue by wojtekmaj** [testing-library/react-testing-library#1197]

**Comment by Lokua** [comment]
> @Lokua You need to set [the `advanceTimers` option](https://testing-library.com/docs/user-event/options#advancetimers) if you're working with fake timers.
> 
> We recommend using a setup function:
> 
> ```js
> function setup(jsx) {
>   return {
>     user: userEvent.setup({
>       advanceTimers: jest.advanceTimersByTime,
>     }),
>     ...render(jsx),
>   }
> }
> 
> test('some click', async () => {
>   jest.useFakeTimers()
>   const { user } = render(<button/>)
> 
>   await user.click(screen.getByRole('button'))
> })
> ```

Thank you very much for the info. I get it now. Just for clarity, in your example, `setup` should actually be called `render`, or vis a versa, yes?

**Comment by ph-fritsche** [comment]
@Lokua Yes, the `setup` function should be called in the test, I fixed the code example above.

**Comment by sangeeth96** [comment]
@ph-fritsche I ran into the same issue but I couldn't get my scenario to work just with `advanceTimers` option. See repro link: https://stackblitz.com/edit/vitest-dev-vitest-2xdjiw?file=test%2Fcountdown.test.tsx

```ts
import Countdown from "src/components/Countdown";
import { render, screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("Mocking timers", () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.runOnlyPendingTimers();
    vi.useRealTimers();
  });

  it("mocks timers", async () => {
    const user = userEvent.setup({
      advanceTimers: ms => vi.advanceTimersByTime(ms),
    });

    render(<Countdown from={5} />);

    await user.click(screen.getByRole("button"));

    expect(screen.getByRole("alert")).toHaveTextContent("5");
    await act(() => vi.runAllTimers());
    expect(screen.getByRole("alert")).toHaveTextContent("0");
  });
});
```

If I do:

```ts
vi.useFakeTimers({ shouldAdvanceTime: true });
```

It makes the test pass but I believe it's not a good option to use.

**Comment by azat-io** [comment]
Same problem

**Comment by bfsgr** [comment]
I believe this might be related to #1187 as vitest uses sinon fake timers

**Comment by xsjcTony** [comment]
Is there any ETA to fix this issue? It should be testing-framework agnostic instead of sticking with `Jest`

One temporarily workaround for `Vitest` users can be like this:

```ts
import { beforeAll, vi } from 'vitest';


// Temporarily workaround for bug in @testing-library/react when use user-event with `vi.useFakeTimers()`
beforeAll(() => {
  const _jest = globalThis.jest;

  globalThis.jest = {
    ...globalThis.jest,
    advanceTimersByTime: vi.advanceTimersByTime.bind(vi)
  };

  return () => void (globalThis.jest = _jest);
});
```

**Comment by eps1lon** [comment]
How did Vitest fake timers ever work before? We generally only support Jest fake timers at the moment. Using any other fake timers means you're on your own.

We'd love to add support for more fake timers but that should come with a general purpose API so that we can support any fake timers (e.g. https://github.com/testing-library/react-testing-library/issues/1187).

**Comment by xsjcTony** [comment]
From what I can see here, `react-testing-library` is hard-coded to use `jest.advanceTimersByTime` after upgrading to `v14`
Current code:
https://github.com/testing-library/react-testing-library/blob/main/src/pure.js#L49-L51
Corresponding PR:
https://github.com/testing-library/react-testing-library/commit/f78839bf4147a777a823e33a429bcf5de9562f9e#diff-2ef28f1bd92d5dcd1f2a04d56814d3adaee10cc939b4a7d7c861af3a3cbbccb7
It's not working like `user-event` where user is allowed to pass their `advanceTimersByTime` functions, like documented here: 
https://testing-library.com/docs/user-event/options#advancetimers

According to this, my workaround above will work 100% perfectly just by binding `jest.advanceTimersByTime` to `vi.advanceTimersByTime`, with binding of `this` to `vi`.

So basically I've no idea why `react-testing-library` just sticks with `Jest`, which is not making sense to me, but as long as I got a stable workaround there so it's all goof for me.

But whatever, thanks for the great testing library!

**Comment by eps1lon** [comment]
> So basically I've no idea why react-testing-library just sticks with Jest

Because it is the most popular testing framework so supporting their fake timers out of the box made sense to help adoption. 

I'd like to support more timers but so far no community contributions have been made to do that. And since I'm not using Vitest in any projects I'm involved in, I didn't have a use-case for myself. PRs are welcome though.

**Comment by wojtekmaj** [comment]
Supporting one framework's fake timers is one thing, making the library framework-dependent by breaking it for all other testing frameworks is another.

**Comment by eps1lon** [comment]
Does it break in vitest with real timers or with their fake timers? I

**Comment by xsjcTony** [comment]
Fake timers

**Comment by iulspop** [comment]
Ran into same problem. Shared my workaround here: https://github.com/testing-library/user-event/issues/1115#issuecomment-1565730917

**Comment by nickmccurdy** [comment]
Supporting Vitest timers should be as easy as calling `vi` instead of `jest`, and historically we've had messier adapter code to deal with differences in DOM implementations, so I think it's worth trying. If anyone's interested I can experiment with a PR.

**Comment by mryechkin** [comment]
@nickmccurdy would absolutely love that! It's literally the only thing stopping me from moving to Vitest 🙏

**Comment by davletovalmir** [comment]
We're using `sinon` fake timers, and we don't have `jest`. For more seamless experience (until it's fully supported), I added the following workaround:

```javascript
import sinon from 'sinon'

const _sinonUseFakeTimers = sinon.useFakeTimers;
let _sinonClock;
sinon.useFakeTimers = function (...args) {
  _sinonClock = _sinonUseFakeTimers.call(sinon, ...args);
  globalThis.jest = { advanceTimersByTime: _sinonClock.tickAsync.bind(_sinonClock) };
  return _sinonClock;
};
```

Might be helpful for those who have similar setup. It allows to focus on the testing, not on hacking things, and can be deleted once sinon timers are supported. Idea was borrowed from https://github.com/testing-library/react-testing-library/issues/1197#issuecomment-1506222246.

**Comment by falldowngoboone** [comment]
> Supporting Vitest timers should be as easy as calling `vi` instead of `jest`, and historically we've had messier code supporting differences in DOM implementations, so I think it's worth trying. If anyone's interested I can experiment with a PR.

It would be great to be able to pass your framework in the `configure` function. Something like:

```javascript
import {configure} from '@testing-library/dom'
import {vi} from 'vitest'

// ...

configure({framework: vi})
```

The object passed through would adhere to a specific interface.

Even if the code has to be messy right now, you will eventually need something that's much more agnostic. Jest won't remain the dominant framework forever.

**Comment by IanVS** [comment]
I found that userEvent can work nicely with vitest fake timers using:

```
    const user = userEvent.setup({
      advanceTimers: ms => vi.advanceTimersByTime(ms),
    });
```

But I didn't find a way to make `waitFor` work correctly.

**Comment by senock-dag** [comment]
Unfortunately the above solution didn't work for me. I've downgraded to `v13.5.0` which has worked.

Is a fix being worked on for future releases?

**Comment by xuhdev** [comment]
@IanVS 's method works for me. Should we add this to the doc?

**Comment by danielholmes** [comment]
The following worked for my case:

```
vi.useFakeTimers({
      shouldAdvanceTime: true
});
```

**Comment by robertjk** [comment]
> The following worked for my case:
> 
> ```
> vi.useFakeTimers({
>       shouldAdvanceTime: true
> });
> ```

I don't think that's a real solution. What `shouldAdvanceTime` option does is advance fake timers with normal time, which kinds of defeat the purpose of using fake timers, and make your test flaky.

**Comment by danielholmes** [comment]
> I don't think that's a real solution. What shouldAdvanceTime option does is advance fake timers with normal time, which kinds of defeat the purpose of using fake timers, and make your test flaky.

I agree it's not a real solution. But if you're dealing with dates in terms of days (my use case), weeks, months or years it's a perfectly fine workaround for now. If your use case needs the precision of milliseconds or seconds then it's not a good workaround.

**Comment by robertjk** [comment]
> > I don't think that's a real solution. What shouldAdvanceTime option does is advance fake timers with normal time, which kinds of defeat the purpose of using fake timers, and make your test flaky.
> 
> I agree it's not a real solution. But if you're dealing with dates in terms of days (my use case), weeks, months or years it's a perfectly fine workaround for now. If your use case needs the precision of milliseconds or seconds then it's not a good workaround.

Fair enough - in some use cases it could be a valid solution.

**Comment by yaroslavvertukhov** [comment]
Any news on solving this problem?

**Comment by yaroslavvertukhov** [comment]
So far my problem has been solved with a crutch

`import { vi } from 'vitest'
globalThis.jest = vi`

**Comment by krutoo** [comment]
when will this problem be solved in this `testing-library/react`?

**Comment by ardunster** [comment]
Just wasted an hour or so fighting with this and had to swap to using `fireEvent` for the clicks in question, which I understand is discouraged. Would be great if this actually gets resolved.
